tc.1
.\"-
.\" Copyright (c) 2022 Cameron Katri.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY CAMERON KATRI AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL CAMERON KATRI OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 19, 2022
.Dt TC 1
.Os
.Sh NAME
.Nm tc
.Nd Create and interact with trustcaches
.Sh SYNOPSIS
.Nm
.Cm append
.Op Fl f Ar flags
.Op Fl u Ar uuid | 0
.Ar infile
.Ar
.Nm
.Cm create
.Op Fl u Ar uuid
.Op Fl v Ar version
.Ar outfile
.Ar
.Nm
.Cm info
.Op Fl c
.Op Fl h
.Op Fl e Ar entrynum
.Ar file
.Nm
.Cm remove
.Op Fl k
.Ar file
.Ar hash ...
.Sh DESCRIPTION
The
.Nm
utility is used to get info about and modify Apple trustcaches.
.Pp
The following commands are supported by
.Nm :
.Bl -tag -width create
.It Xo
.Cm append
.Op Fl f Ar flags
.Op Fl u Ar uuid | 0
.Ar infile
.Ar
.Xc
Modify the trustcache at
.Ar infile
to include each signed Mach-O at the specified paths.
If
.Ar file
is both 40 characters and hexadecimal, that hash will be added to the cache.
.Ar uuid
is used to specify a custom uuid to be used.
If it is
.Ar 0 ,
the uuid will be left the same, otherwise, it will be regenerated.
If
.Fl f
is specified, any new entries with have the flags specified at
.Ar flags .
.It Xo
.Cm create
.Op Fl u Ar uuid
.Op Fl v Ar version
.Ar outfile
.Ar
.Xc
Create a trustcache at
.Ar outfile .
Each Mach-O found in the specified inputs will be scanned for
a code signature and hashed.
Any malformed or unsigned Mach-O will be ignored.
Each slice of a FAT binary will have its hash included.
Versions 0 and 1 are supported, if not specified, 1 is assumed.
If
.Ar uuid
is specified, that will be used instead of a randomly generated one.
.It Xo
.Cm info
.Op Fl c
.Op Fl h
.Op Fl e Ar entrynum
.Ar file
.Xc
Print information about
.Ar file .
The output for each hash will be in the format:
.Pp
.Dl <cdhash> <flags> [<hash_type>]
.Pp
If the
.Fl c
is given, only the hashes will be printed.
If
.Fl h
is given, only the header will be printed.
If
.Ar entrynum
is specified, only that entry will be printed.
.It Xo
.Cm remove
.Op Fl k
.Ar file
.Ar hash ...
.Xc
Remove each specified hash from
.Ar file .
If
.Fl k
is specified, the uuid will not be regenerated.
The number of removed entries will be printed.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr cryptex-dump-trust-cache 1 ,
.Xr cryptex-generate-trust-cache 1
.Sh HISTORY
The
.Nm
utility was written by
.An Cameron Katri Aq Mt me@cameronkatri.com .


tc.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2022 Cameron Katri.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY CAMERON KATRI AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL CAMERON KATRI OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>

#include "trustcache.h"

int
main(int argc, char **argv)
{
	if (argc < 2) {
help:
		fprintf(stderr, "Usage: tc append [-f flags] [-u uuid | 0] infile file ...\n"
										"       tc create [-u uuid] [-v version] outfile file ...\n"
										"       tc info [-c] [-h] [-e entrynum] file\n"
										"       tc remove [-k] file hash ...\n\n"
										"See tc(1) for more information\n");
		exit(1);
	}
	
	int ret = 1;

	if (strcmp(argv[1], "info") == 0)
		ret = tcinfo(argc - 1, argv + 1);
	else if (strcmp(argv[1], "create") == 0)
		ret = tccreate(argc - 1, argv + 1);
	else if (strcmp(argv[1], "append") == 0)
		ret = tcappend(argc - 1, argv + 1);
	else if (strcmp(argv[1], "remove") == 0)
		ret = tcremove(argc - 1, argv + 1);
	else
		fprintf(stderr, "Unknown subcommand %s\n", argv[1]);

	if (ret == -1)
		goto help;

	return ret;
}

struct trust_cache
opentrustcache(const char *path)
{
	FILE *f;
	struct trust_cache cache;

	if ((f = fopen(path, "r")) == NULL) {
		fprintf(stderr, "%s: %s\n", path, strerror(errno));
		exit(1);
	}

	fread(&cache, sizeof(struct trust_cache) - sizeof(struct trust_cache_entry1*), 1, f);
	cache.version = le32toh(cache.version);
	cache.num_entries = le32toh(cache.num_entries);

	if (cache.version == 0) {
		if ((cache.hashes = calloc(cache.num_entries, sizeof(trust_cache_hash0))) == NULL)
			exit(EX_OSERR);
		fread(cache.hashes, sizeof(trust_cache_hash0), cache.num_entries, f);
	} else if (cache.version == 1) {
		if ((cache.entries = calloc(cache.num_entries, sizeof(struct trust_cache_entry1))) == NULL)
			exit(EX_OSERR);
		fread(cache.entries, sizeof(struct trust_cache_entry1), cache.num_entries, f);
	} else {
		fprintf(stderr, "%s: Unsupported version %i\n", path, cache.version);
		exit(1);
	}

	fclose(f);
	return cache;
}
